pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,430 @@
+package tfbridge
+
+import (
+	"context"
+	"errors"
+	"slices"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+)
+
+func isPresent(val resource.PropertyValue) bool {
+	return !val.IsNull() &&
+		!(val.IsArray() && val.ArrayValue() == nil) &&
+		!(val.IsObject() && val.ObjectValue() == nil)
+}
+
+func isForceNew(tfs shim.Schema, ps *SchemaInfo) bool {
+	if tfs != nil && tfs.ForceNew() {
+		return true
+	}
+	if ps != nil && ps.ForceNew != nil && *ps.ForceNew {
+		return true
+	}
+	return false
+}
+
+func isObject(tfs shim.Schema, ps *SchemaInfo) bool {
+	if tfs.Type() == shim.TypeMap {
+		return true
+	}
+
+	if tfs.Type() != shim.TypeList && tfs.Type() != shim.TypeSet {
+		return false
+	}
+
+	if ps != nil && ps.MaxItemsOne != nil {
+		return *ps.MaxItemsOne
+	}
+
+	return tfs.MaxItems() == 1
+}
+
+func sortedMergedKeys(a resource.PropertyMap, b resource.PropertyMap) []resource.PropertyKey {
+	keys := make(map[resource.PropertyKey]struct{})
+	for k := range a {
+		keys[k] = struct{}{}
+	}
+	for k := range b {
+		keys[k] = struct{}{}
+	}
+	keysSlice := make([]resource.PropertyKey, 0, len(keys))
+	for k := range keys {
+		keysSlice = append(keysSlice, k)
+	}
+	slices.Sort(keysSlice)
+	return keysSlice
+}
+
+func promoteToReplace(diff *pulumirpc.PropertyDiff) *pulumirpc.PropertyDiff {
+	kind := diff.GetKind()
+	switch kind {
+	case pulumirpc.PropertyDiff_ADD:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE}
+	case pulumirpc.PropertyDiff_DELETE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE}
+	case pulumirpc.PropertyDiff_UPDATE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE}
+	default:
+		return diff
+	}
+}
+
+type baseDiff string
+
+const (
+	NoDiff    baseDiff = "NoDiff"
+	Add       baseDiff = "Add"
+	Delete    baseDiff = "Delete"
+	Update    baseDiff = "Update"
+	Undecided baseDiff = "Undecided"
```

Comment:
These constants don't need to be public:
```suggestion
	noDiff    baseDiff = "NoDiff"
	add       baseDiff = "Add"
	delete    baseDiff = "Delete"
	update    baseDiff = "Update"
	undecided baseDiff = "Undecided"
```